nixos/luksroot: Add `askForPin` option

jmbaur opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

https://discourse.nixos.org/t/using-fido2-luks-with-yubikey-pin/15484
With a Yubikey configured to require a PIN for Fido2, the current fido2luks setup does not work without using the `--pin` and `--pin-source` flags.

###### Things done

Added shell code to (if option enabled) request user for device PIN. Write that PIN to /crypt-ramfs/passphrase. Use the PIN with `fido2luks`. Then remove /crypt-ramfs/passphrase. Also add relevant nixos module option under `boot.initrd.luks.devices.<name>.fido2.askForPin`.

Tested on my personal machine :)

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jmbaur committed:
Add --pin for testing with fido2luks

jmbaur committed:
Add pin source

jmbaur committed:
Read pin directly

jmbaur committed:
Use --pin flag

jmbaur committed:
Add helpful message

jmbaur committed:
Hide extra shell code behind nix module option

jmbaur committed:
Add setup description to docs section

jmbaur commented:
I am unsure why there are so many whitespace changes, I just used `nixpkgs-fmt` with the relevant files.

Artturin commented:
the file wasn't formatted with nixpkgs-fmt to begin with (most files aren't... yet) could you please revert the formatting changes so that this will be easier to review?

jmbaur committed:
Use <literal></literal> instead of backticks

Artturin commented on nixos/doc/manual/configuration/luks-file-systems.section.md:
````diff
@@ -75,3 +75,10 @@ as [Trezor](https://trezor.io/).
 boot.initrd.luks.devices."/dev/sda2".fido2.passwordLess = true;
 ```
+
+With a passwordless setup described above, you can enable `askForPin` if your
````
remember to run `./nixos/doc/manual/md-to-db.sh` after making changes to the manual

jmbaur replied:
Just did, thanks for the reminder!

jmbaur committed:
Add from_md docs

jmbaur committed:
Revert whitespace changes

jmbaur commented:
> the file wasn't formatted with nixpkgs-fmt to begin with (most files aren't... yet) could you please revert the formatting changes so that this will be easier to review?

Whitespace changes are reverted!

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

Artturin commented:
found an existing issue for this https://github.com/NixOS/nixpkgs/issues/141824

and a existing pr https://github.com/NixOS/nixpkgs/pull/143981
please take a look and discuss and decide which one of the approaches is better

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

peterwilli commented:
> found an existing issue for this #141824
> 
> and a existing pr #143981 please take a look and discuss and decide which one of the approaches is better

I made #143981 but I'll try to be as objective as I can:

Main advantages of this PR:

 - The pin/passphrase can be reused across different LUKS devices as it resides in RAM using RAMFS for the lifecycle of the boot sequnce.


Disadvantages:
 - It has to be stressed that other processes and applications can still access this file, and it we can't track which processes accesses it and when. The file also has to be **carefully shredded** after the boot has been completed (which this PR is not doing correctly, it should use `shred` rather than `rm` to make sure it is really vanished.

Main advantages #143981:

 - Named pipe also resides in RAM, but is directly sent to the decryptor process (in our case, `fido2luks`), after the data has arrived at the process, it is gone. There's no need to worry about security, as as long as `fido2luks` got your key, you know you're safe and no one else has it. Accidental wrong handling of the named pipe will not result in the pin or passphrase still residing in RAM
 
Disadvantages:

 - Using `fido2luks` in NixOS twice will result in being asked for the same pin twice. This is something I haven't thought about as I only open 1 drive at a time at boot, I have more than 1 LUKs drive, but open those after I booted in the desktop.

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

jmbaur committed:
Don't remove pin from ramfs

Keep Fido2 device's pin in ramfs for a potential usage of the pin with
other luks devices (e.g. when `boot.initrd.luks.reusePassphrases` is
used.

jmbaur committed:
Merge remote-tracking branch 'origin/master' into fido2luks-pin

jmbaur committed:
Merge branch 'fido2luks-pin' of ssh://github.com/jmbaur/nixpkgs into fido2luks-pin

jmbaur commented:
@peterwilli I do like your proposed solution. There are other uses of `rm -f /crypt-ramfs/passphrase` throughout the luksroot file, including in the functions `do_open_gpg_card`, `do_open_yubikey`, and `do_open_passphrase`. Are you proposing they would also potentially need to be changed?

jmbaur committed:
Remove Fido2 pin if not going to be reused

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

jmbaur committed:
Merge branch 'fido2luks-pin' of ssh://github.com/jmbaur/nixpkgs into fido2luks-pin

jmbaur committed:
Syntax fix

peterwilli commented:
> @peterwilli I do like your proposed solution. There are other uses of `rm -f /crypt-ramfs/passphrase` throughout the luksroot file, including in the functions `do_open_gpg_card`, `do_open_yubikey`, and `do_open_passphrase`. Are you proposing they would also potentially need to be changed? If so, should that be handled in the separate PR so the implementation remains consistent with other LUKS device opening strategies?

While I'm not really familiar with crypt-ramfs, I know that `rm` doesn't really delete a file but marks it as deleted until something else is being stores, overwriting that free space. It doesn't harm anything, and if there's a chance something can read memory aside from the intended processes then `shred` will definitely prevent that.

`shred` basically writes the bytes that previously occupied the file with random data so that it's harder to read back. In my opinion, especially since it's a drop-in replacement for `rm` (at least in this case), we should go with that.

It seems your pull request is the best to go with as it makes sense to keep a single implementation to keep things simple and consistent when it comes to cryptography. It seems to be the best to go with yours and replace instances of rm with `shred`.

I feel kinda bad for not realizing I should've gone with the files approach as well (since it was everywhere already), especially since I spent so much time on getting the package to work. But in the end we're all learning.

peterwilli committed:
luksroot: replaced rm with shred

peterwilli commented:
Hi @jmbaur. I tested with `shred` and it works great under your pull request. I sent you a PR to have this change in your PR: https://github.com/jmbaur/nixpkgs/pull/1

After that, I'd say we should go with this PR as our fido2luks solution!

jmbaur committed:
Merge pull request #1 from peterwilli/fido2luks-pin

luksroot: replaced rm with shred

jmbaur commented:
@peterwilli Merged in your changes and they work great! Thanks!

peterwilli commented:
That's great to hear! I'll close my PR, this is much more consistent

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

ck3d commented on nixos/modules/system/boot/luksroot.nix:
```diff
@@ -432,11 +432,39 @@ let
+                    if [ -n "$pin" ]; then
+                       ${if luks.reusePassphrases then ''
+                         # Remember it for the next device
+                         echo -n "$pin" > /crypt-ramfs/passphrase
```
I think the code only works if reusePassphrases is set to true. If reusePassphrases is set to false, /crypt-ramfs/passphrase is never written and fido2luks --pin-source will point to a not existing file.

jmbaur replied:
Good catch. I pushed up a change that fixes this.

jmbaur committed:
Merge branch 'NixOS:master' into fido2luks-pin

jmbaur committed:
Save to `/crypt-ramfs/passphrase` always

After using the pin from `/crypt-ramfs/passphrase`, if
`luks.reusePassphrases` is false, then delete it, otherwise it stays.

peterwilli commented on nixos/modules/system/boot/luksroot.nix:
```diff
@@ -444,12 +444,7 @@ let
-                         # Remember it for the next device
-                         echo -n "$pin" > /crypt-ramfs/passphrase
-                       '' else ''
-                         # Don't save it to ramfs. We are very paranoid
```
Why do we always store to crypt-ramfs? I was just curious.

jmbaur replied:
I'm sure there are many other reasons I am not familiar with, but I do know that the `/crypt-ramfs/passphrase` path is used for reuse of a key to decrypt multiple drives (configured via `boot.initrd.luks.reusePassphrases`.

peterwilli replied:
Oh, so not writing to it even when reuse not being enabled could break behavior, you say?

jmbaur replied:
Actually, not it does not! This could use bash process substitution to read from an environment variable so that `--pin-source` will work. From the behavior I have seen, `--pin-source` requires the source to be a file. If process substitution is used, the pin doesn't have to be written to `/crypt-ramfs/...` and then conditionally removed.

jmbaur commented:
I'm not sure if renaming `reusePassphrases` would be appropriate since it is applicable to more than just unlocking a disk with the help of a FIDO2 device, where `reusePin` may make more sense. You do make a good point about the salt as well, that the entire passphrase consists of an optional salt, pin if required by the device, and what the FIDO2 device provides, so reusing a passphrase in this context could mean saving the pin and salt for other disks to use. Potentially, something like `fido2luks print-secret` could be used to just save the final secret for unlocking the disk, but I believe that would mean the same `fido2luks` credential would need to be used for each disk (is that an anti-pattern?).

peterwilli commented:
> I'm not sure if renaming `reusePassphrases` would be appropriate since it is applicable to more than just unlocking a disk with the help of a FIDO2 device, where `reusePin` may make more sense. You do make a good point about the salt as well, that the entire passphrase consists of an optional salt, pin if required by the device, and what the FIDO2 device provides, so reusing a passphrase in this context could mean saving the pin and salt for other disks to use. Potentially, something like `fido2luks print-secret` could be used to just save the final secret for unlocking the disk, but I believe that would mean the same `fido2luks` credential would need to be used for each disk (is that an anti-pattern?).

I think that indeed is an anti-pattern, because we shouldn't assume one setup over the other. The way I understand it, is it'd mean that both the salt and the pin should be equal for both disks or none at all.

It's kind of hard to tell really, like, if one really wants both disks to be unlockable using a different salt or pin, then it's doubtful you wish to unlock it at boot-time. It *could* be safe to say that if you wish to decrypt disks at boot, they share the same key, whether it be a pin or salt.

But I could be wrong there. I just don't really see the appeal nor the gain in security in typing 2 different salts at boot time, each boot. 

Maybe someone who has a setup like this can step in and give a reason why they have it!

Until then, I think that at least allowing the user to reuse the salt in addition to the pin makes sense, since it's more likely they have the same salt if they also use the same pin.

As for to rename `reusePassphrases`, I had another reason for this. When we boot into the bootloader we ask the user to enter a passphrase (not part of our PR) (see: https://github.com/NixOS/nixpkgs/blob/9d5ebae270cc8d59ac9bc797a2cc651ea58307e0/nixos/modules/system/boot/luksroot.nix#L163).

This is actually the salt, and not really the passphrase to actually unlock a LUKS drive. If we keep `reusePassphrases`, it maybe makes sense to change the prompt for when FIDO2LUKS in enabled, because at one point or other there is some kind of conflict in terminology that I think has to be resolved if we offer the user more options to reuse their keys (whatever they might be).

jmbaur commented:
Or `reusePassphrases` could just mean save the salt and/or pin for use with future disks, then an extra option would not have to be introduced. Thoughts?

peterwilli commented:
> Or `reusePassphrases` could just mean save the salt and/or pin for use with future disks, then an extra option would not have to be introduced. Thoughts?

What do you mean by "and/or"? How can users decide if they want to store the salt or the pin, or both?

smkuehnhold commented:
(Sorry if inappropriate/off-topic)

@jmbaur @peterwilli 

What is the status of this PR? Is the lack of progress in the last 4 months due simply to fading interest or some other concern? 

I don't have the knowledge myself to contribute in a reasonable time frame, but I am curious if it would be worth my time to at least dip my toes in the water 😅.

jmbaur commented:
This PR allows for unlocking with a hardware FIDO key that requires a pin, but AFAIK only works with one disk. @peterwilli brought up the use case for unlocking multiple disks, which adds complexity to the unlocking script. I believe it may be a better idea to let systemd handle all of this in initrd. I know systemd in initrd for NixOS is quite new though and not fully functional (see PR https://github.com/NixOS/nixpkgs/pull/179823).

peterwilli commented:
> (Sorry if inappropriate/off-topic)
> 
> @jmbaur @peterwilli
> 
> What is the status of this PR? Is the lack of progress in the last 4 months due simply to fading interest or some other concern?
> 
> I don't have the knowledge myself to contribute in a reasonable time frame, but I am curious if it would be worth my time to at least dip my toes in the water sweat_smile.

No lack of interest, it is just ready and it isn't getting merged 😂 I been using it since god knows how long.
The main problem now is that my work involves:

 - Pulling the latest `nixpkgs`
 - Apply the patch
 - `nixos-rebuild`

I don't know what to do to make it easier to use without it being merged.

smkuehnhold commented:
@peterwilli 

Is possible to simply [replace the module via nixos config](https://nixos.org/manual/nixos/stable/#sec-replace-modules)? Or does something about this patch make it untenable? 

I coincidentally came across this feature while perusing manual earlier today,  so I'd thought I'd ask 🤷🏼‍♂️

peterwilli commented:
Hey everyone, through the NixOS Discord NobbZ showed me a way to handle this easier while not having to keep updating your own `nixpkgs`-fork.

I followed this tutorial: https://nixos.org/manual/nixos/stable/index.html#sec-replace-modules
My demo config is here: https://keybase.pub/peterwilli/NixOS%20Shared%20Stuff/desktop/configuration.nix

I'm currently rebuilding it'll take a while, but I think it'll work.

jmbaur committed:
Merge branch 'master' into fido2luks-pin

jmbaur commented:
I have no personal interest in using the changes here. I'll close this. Feel free to continue if someone else finds use in it!
